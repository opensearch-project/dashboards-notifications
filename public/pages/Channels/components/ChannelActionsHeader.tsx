/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiButton, EuiContextMenuItem, EuiPopover } from '@elastic/eui';
import React, { useState } from 'react';
import { ChannelItemType } from '../../../../models/interfaces';
import { ModalConsumer } from '../../../components/Modal';
import { ROUTES } from '../../../utils/constants';
import { DeleteChannelModal } from './modals/DeleteChannelModal';

interface ChannelActionsParams {
  label: string;
  disabled: boolean;
  modal?: React.ReactNode;
  modalParams?: object;
  href?: string;
  action?: () => void;
}

interface ChannelActionsProps {
  selected: ChannelItemType[];
  setSelected: (items: ChannelItemType[]) => void;
  items: ChannelItemType[];
  setItems: (items: ChannelItemType[]) => void;
  refresh: () => void;
  href?: string;
}

export function ChannelActionsHeader(props: ChannelActionsProps) {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);

  const actions: ChannelActionsParams[] = [
    {
      label: 'Edit',
      disabled: props.selected.length !== 1,
      href: props.href,
    },
    {
      label: 'Delete',
      disabled: props.selected.length === 0,
      modal: DeleteChannelModal,
      modalParams: { refresh: props.refresh },
    },
  ];

  return (
    <ModalConsumer>
      {({ onShow }) => (
        <EuiPopover
          panelPaddingSize="none"
          button={
            <EuiButton
              iconType="arrowDown"
              iconSide="right"
              disabled={props.selected.length === 0}
              onClick={() => setIsPopoverOpen(!isPopoverOpen)}
            >
              Actions
            </EuiButton>
          }
          isOpen={isPopoverOpen}
          closePopover={() => setIsPopoverOpen(false)}
        >
          {actions.map((params) => (
            <EuiContextMenuItem
              key={params.label}
              disabled={params.disabled}
              onClick={() => {
                setIsPopoverOpen(false);
                if (params.modal) {
                  onShow(params.modal, {
                    selected: props.selected,
                    ...(params.modalParams || {}),
                  });
                }
                if (params.href) location.assign(params.href);
                if (params.action) params.action();
              }}
            >
              {params.label}
            </EuiContextMenuItem>
          ))}
        </EuiPopover>
      )}
    </ModalConsumer>
  );
}
