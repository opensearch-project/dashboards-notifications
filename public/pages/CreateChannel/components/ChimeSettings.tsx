/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiCompressedFieldText, EuiCompressedFormRow } from '@elastic/eui';
import React, { useContext } from 'react';
import { CreateChannelContext } from '../CreateChannel';
import { validateWebhookURL } from '../utils/validationHelper';

interface ChimeSettingsProps {
  chimeWebhook: string;
  setChimeWebhook: (url: string) => void;
}

export function ChimeSettings(props: ChimeSettingsProps) {
  const context = useContext(CreateChannelContext)!;

  return (
    <EuiCompressedFormRow
      label="Webhook URL"
      style={{ maxWidth: '700px' }}
      error={context.inputErrors.chimeWebhook.join(' ')}
      isInvalid={context.inputErrors.chimeWebhook.length > 0}
    >
      <EuiCompressedFieldText
        fullWidth
        data-test-subj="create-channel-chime-webhook-input"
        placeholder="https://hooks.chime.aws/incomingwebhooks/XXXXX..."
        value={props.chimeWebhook}
        onChange={(e) => props.setChimeWebhook(e.target.value)}
        isInvalid={context.inputErrors.chimeWebhook.length > 0}
        onBlur={() => {
          context.setInputErrors({
            ...context.inputErrors,
            chimeWebhook: validateWebhookURL(props.chimeWebhook),
          });
        }}
      />
    </EuiCompressedFormRow>
  );
}
