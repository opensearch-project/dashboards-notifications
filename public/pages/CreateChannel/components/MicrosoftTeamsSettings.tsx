/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiFieldText, EuiFormRow } from '@elastic/eui';
import React, { useContext } from 'react';
import { CreateChannelContext } from '../CreateChannel';
import { validateWebhookURL } from '../utils/validationHelper';

interface MicrosoftTeamsSettingsProps {
  microsoftTeamsWebhook: string;
  setMicrosoftTeamsWebhook: (url: string) => void;
}

export function MicrosoftTeamsSettings(props: MicrosoftTeamsSettingsProps) {
  const context = useContext(CreateChannelContext)!;

  return (
    <EuiFormRow
      label="Webhook URL"
      style={{ maxWidth: '700px' }}
      error={context.inputErrors.microsoftTeamsWebhook.join(' ')}
      isInvalid={context.inputErrors.microsoftTeamsWebhook.length > 0}
    >
      <EuiFieldText
        fullWidth
        data-test-subj="create-channel-microsoftTeams-webhook-input"
        placeholder="https://xxxxx.webhook.office.com/webhookb2/xxxxx/IncomingWebhook/xxxxx..."
        value={props.microsoftTeamsWebhook}
        onChange={(e) => props.setMicrosoftTeamsWebhook(e.target.value)}
        isInvalid={context.inputErrors.microsoftTeamsWebhook.length > 0}
        onBlur={() => {
          context.setInputErrors({
            ...context.inputErrors,
            microsoftTeamsWebhook: validateWebhookURL(props.microsoftTeamsWebhook),
          });
        }}
      />
    </EuiFormRow>
  );
}
