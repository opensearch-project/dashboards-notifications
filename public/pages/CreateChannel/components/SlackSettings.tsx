/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiCompressedFieldText, EuiCompressedFormRow } from '@elastic/eui';
import React, { useContext } from 'react';
import { CreateChannelContext } from '../CreateChannel';
import { validateWebhookURL } from '../utils/validationHelper';

interface SlackSettingsProps {
  slackWebhook: string;
  setSlackWebhook: (url: string) => void;
}

export function SlackSettings(props: SlackSettingsProps) {
  const context = useContext(CreateChannelContext)!;

  return (
    <EuiCompressedFormRow
      label="Slack webhook URL"
      style={{ maxWidth: '700px' }}
      error={context.inputErrors.slackWebhook.join(' ')}
      isInvalid={context.inputErrors.slackWebhook.length > 0}
    >
      <EuiCompressedFieldText
        fullWidth
        data-test-subj="create-channel-slack-webhook-input"
        placeholder="https://hooks.slack.com/services/XXXXX/XXXXX/XXXXX..."
        value={props.slackWebhook}
        onChange={(e) => props.setSlackWebhook(e.target.value)}
        isInvalid={context.inputErrors.slackWebhook.length > 0}
        onBlur={() => {
          context.setInputErrors({
            ...context.inputErrors,
            slackWebhook: validateWebhookURL(props.slackWebhook),
          });
        }}
      />
    </EuiCompressedFormRow>
  );
}
