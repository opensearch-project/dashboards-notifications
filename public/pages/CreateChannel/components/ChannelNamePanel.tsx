/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiFieldText, EuiFormRow, EuiTextArea } from '@elastic/eui';
import React, { useContext } from 'react';
import { ContentPanel } from '../../../components/ContentPanel';
import { CreateChannelContext } from '../CreateChannel';
import { validateChannelName } from '../utils/validationHelper';
import { i18n } from '@osd/i18n';

interface ChannelNamePanelProps {
  name: string;
  setName: (name: string) => void;
  description: string;
  setDescription: (description: string) => void;
}

export function ChannelNamePanel(props: ChannelNamePanelProps) {
  const context = useContext(CreateChannelContext)!;
  return (
    <>
      <ContentPanel
        bodyStyles={{ padding: 'initial' }}
        title={
          i18n.translate('notification.notificationChannels.nameAndDescription', {
          defaultMessage:
          "Name and description",
          })}
        
        titleSize="s"
      >
        <EuiFormRow
          label={
            i18n.translate('notification.notificationChannels.newChannelName', {
            defaultMessage:
              'Name',
            })}
          
          error={context.inputErrors.name.join(' ')}
          isInvalid={context.inputErrors.name.length > 0}
        >
          <EuiFieldText
            data-test-subj="create-channel-name-input"
            placeholder={
              i18n.translate('notification.notificationChannels.newChannelPlaceholder', {
              defaultMessage:
                "Enter channel name",
              })}
            
            value={props.name}
            onChange={(e) => props.setName(e.target.value)}
            isInvalid={context.inputErrors.name.length > 0}
            onBlur={() => {
              context.setInputErrors({
                ...context.inputErrors,
                name: validateChannelName(props.name),
              });
            }}
          />
        </EuiFormRow>
        <EuiFormRow
          label={
            <span>
              {
              i18n.translate('notification.notificationChannels.newChannelDescription', {
              defaultMessage:
                "Description - ",
              })}
              <i style={{ fontWeight: 'normal' }}>{
                i18n.translate('notification.notificationChannels.newChannelDescriptionOptional', {
                defaultMessage:
                  'optional',
                })}
                </i>
            </span>
          }
        >
          <>
            <EuiTextArea
              data-test-subj="create-channel-description-input"
              placeholder={
                i18n.translate('notification.notificationChannels.newChannelPurpose', {
                defaultMessage:
                "What is the purpose of this channel?",
                })}
              
              style={{ height: '4.1rem' }}
              value={props.description}
              onChange={(e) => props.setDescription(e.target.value)}
            />
          </>
        </EuiFormRow>
      </ContentPanel>
    </>
  );
}
