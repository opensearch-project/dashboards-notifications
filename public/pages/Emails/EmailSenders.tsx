/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiSpacer, EuiTitle } from '@elastic/eui';
import React, { useContext, useEffect } from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { CoreServicesContext } from '../../components/coreServices';
import { BREADCRUMBS } from '../../utils/constants';
import { MainContext } from '../Main/Main';
import { SendersTable } from './components/tables/SendersTable';
import { SESSendersTable } from './components/tables/SESSendersTable';
import { NotificationService } from '../../services';
import { i18n } from '@osd/i18n';

interface EmailSendersProps extends RouteComponentProps {
  notificationService: NotificationService;
}

export function EmailSenders(props: EmailSendersProps) {
  const coreContext = useContext(CoreServicesContext)!;
  const mainStateContext = useContext(MainContext)!;


  useEffect(() => {
    coreContext.chrome.setBreadcrumbs([
      BREADCRUMBS.NOTIFICATIONS,
      BREADCRUMBS.EMAIL_SENDERS,
    ]);
    window.scrollTo(0, 0);
  }, []);

  return (
    <>
      <EuiTitle size="l">
      <h1>
        {
          i18n.translate('notification.notificationChannels.emailSenders', {
          defaultMessage:
          'Email senders',
        })}
        </h1>
      </EuiTitle>
      {mainStateContext.availableConfigTypes.includes('smtp_account') && (
        <>
          <EuiSpacer />
          <SendersTable coreContext={coreContext} notificationService={props.notificationService} />
        </>
      )}

      {/* UI currently does not fully handle this condition, adding it just to avoid flashing */}
      {mainStateContext.availableConfigTypes.includes('ses_account') && (
        <>
          <EuiSpacer />
          <SESSendersTable coreContext={coreContext} notificationService={props.notificationService} />
        </>
      )}
    </>
  );
}
