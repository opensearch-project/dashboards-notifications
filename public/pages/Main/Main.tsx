/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiPage, EuiPageBody, EuiPageSideBar, EuiSideNav } from '@elastic/eui';
import React, { Component, createContext, useContext } from 'react';
import { Redirect, Route, RouteComponentProps, Switch } from 'react-router-dom';
import { CoreStart, SavedObject } from '../../../../../src/core/public';
import { CoreServicesConsumer, CoreServicesContext } from '../../components/coreServices';
import { ModalProvider, ModalRoot } from '../../components/Modal';
import { BrowserServices } from '../../models/interfaces';
import { ServicesConsumer, ServicesContext } from '../../services/services';
import { ROUTES } from '../../utils/constants';
import { CHANNEL_TYPE } from '../../../common/constants';
import { Channels } from '../Channels/Channels';
import { ChannelDetails } from '../Channels/components/details/ChannelDetails';
import { CreateChannel } from '../CreateChannel/CreateChannel';
import { CreateRecipientGroup } from '../Emails/CreateRecipientGroup';
import { CreateSender } from '../Emails/CreateSender';
import { CreateSESSender } from '../Emails/CreateSESSender';
import { EmailGroups } from '../Emails/EmailGroups';
import { EmailSenders } from '../Emails/EmailSenders';
import { DataSourceMenuContext, DataSourceMenuProperties } from "../../services/DataSourceMenuContext";
import queryString from "query-string";
import {
  DataSourceManagementPluginSetup,
  DataSourceSelectableConfig,
  DataSourceViewConfig,
} from "../../../../../src/plugins/data_source_management/public";
import { DataSourceOption } from "../../../../../src/plugins/data_source_management/public/components/data_source_menu/types";
import _ from "lodash";
import { NotificationService } from '../../services';
import { HttpSetup } from '../../../../../src/core/public';
import * as pluginManifest from "../../../opensearch_dashboards.json";
import { DataSourceAttributes } from "../../../../../src/plugins/data_source/common/data_sources";
import semver from "semver";

import * as http from 'http';
import { i18n } from '@osd/i18n';


enum Navigation {
  Notifications = i18n.translate('notification.notificationChannels.notificationsTitle', {
    defaultMessage:
      'Notifications',
    }),
  Channels = i18n.translate('notification.notificationChannels.', {
    defaultMessage:
    i18n.translate('notification.notificationChannels.channelsList', {
      defaultMessage:
      'Channels',
      }), 
    }),
  EmailSenders = i18n.translate('notification.notificationChannels.emailSenders', {
    defaultMessage:
    'Email senders',
    }),
  EmailGroups = i18n.translate('notification.notificationChannels.emailRecipientsGroupsList', {
    defaultMessage:
    'Email recipient groups',
    }),
}

enum Pathname {
  Channels = '/channels',
}

interface MainProps extends RouteComponentProps {
  setActionMenu: (menuMount: MountPoint | undefined) => void;
  multiDataSourceEnabled: boolean;
  dataSourceManagement: DataSourceManagementPluginSetup;
  defaultRoute?: string;
}

export interface MainState extends Pick<DataSourceMenuProperties, "dataSourceId" | "dataSourceLabel"> {
  availableChannels: Partial<typeof CHANNEL_TYPE>;
  availableConfigTypes: string[]; // available backend config types
  tooltipSupport: boolean; // if true, IAM role for SNS is optional and helper text should be available
  dataSourceReadOnly: boolean;
  dataSourceLoading: boolean;
  dataSourceLabel: string;
}

export const MainContext = createContext<MainState | null>(null);

export default class Main extends Component<MainProps, MainState> {
  static contextType = ServicesContext;
  constructor(props: MainProps) {
    super(props);
    const initialState = {
      availableChannels: CHANNEL_TYPE,
      availableConfigTypes: [],
      tooltipSupport: false,
    };

    if (props.multiDataSourceEnabled) {
      const {
        dataSourceId = "",
        dataSourceLabel = ""
      } = queryString.parse(this.props.location.search) as {
        dataSourceId?: string;
        dataSourceLabel?: string;
      };

      this.state = {
        ...initialState,
        dataSourceId: dataSourceId,
        dataSourceLabel: dataSourceLabel,
        dataSourceReadOnly: false,
        dataSourceLoading: props.multiDataSourceEnabled,
      };
    } else {
      this.state = initialState;
    }
  }

  async componentDidMount() {
    this.setServerFeatures();
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.props.multiDataSourceEnabled && (prevState.dataSourceId !== this.state.dataSourceId)) {
      // Call setServerFeatures when dataSourceId is updated or dataSourceComponent is loaded
      this.setServerFeatures();
    }
  }

  async setServerFeatures(): Promise<void> {
    const services = this.getServices(this.props.http);
    const serverFeatures = await services.notificationService.getServerFeatures();
    const defaultConfigTypes = [
      'slack',
      'chime',
      'microsoft_teams',
      'webhook',
      'email',
      'sns',
      'smtp_account',
      'ses_account',
      'email_group',
    ];

    let newState = {
      dataSourceId: this.state.dataSourceId || '',
      dataSourceLabel: this.state.dataSourceLabel || '',
      dataSourceReadOnly: false,
      dataSourceLoading: this.state.dataSourceLoading,
      availableChannels: this.props.multiDataSourceEnabled ? CHANNEL_TYPE : defaultConfigTypes,
      availableConfigTypes: defaultConfigTypes,
      tooltipSupport: false,
    };

    if (serverFeatures) {
      const { availableChannels, availableConfigTypes, tooltipSupport } = serverFeatures;
      newState = {
        ...newState,
        availableChannels,
        availableConfigTypes,
        tooltipSupport,
      };
    }

    this.setState(newState);
  }

  onSelectedDataSources = (dataSources: DataSourceOption[]) => {
    const { id = "", label = "" } = dataSources[0] || {};
    if (this.state.dataSourceId !== id || this.state.dataSourceLabel !== label) {
      this.setState({
        dataSourceId: id,
        dataSourceLabel: label,
      });
    }
    if (this.state.dataSourceLoading) {
      this.setState({
        dataSourceLoading: false,
      });
    }
  };

  dataSourceFilterFn = (dataSource: SavedObject<DataSourceAttributes>) => {
    const dataSourceVersion = dataSource?.attributes?.dataSourceVersion || "";
    const installedPlugins = dataSource?.attributes?.installedPlugins || [];
    return (
      semver.satisfies(dataSourceVersion, pluginManifest.supportedOSDataSourceVersions) &&
      pluginManifest.requiredOSDataSourcePlugins.every((plugin) => installedPlugins.includes(plugin))
    );
  };

  getServices(http: HttpSetup) {
    const {
      location: { pathname },
    } = this.props;
    let notificationService;
    if (this.props.multiDataSourceEnabled) {
      notificationService = new NotificationService(http, this.state.dataSourceId, this.props.multiDataSourceEnabled);
    }
    else {
      notificationService = new NotificationService(http);
    }
    const services = {
      notificationService,
    };
    return services;
  }

  render() {
    const {
      location: { pathname },
    } = this.props;
    let DataSourceMenuSelectable, DataSourceMenuView;
    let activeOption: DataSourceOption[] | undefined;
    if (this.props.multiDataSourceEnabled) {
      DataSourceMenuSelectable = this.props.dataSourceManagement?.ui?.getDataSourceMenu<DataSourceSelectableConfig>();
      DataSourceMenuView = this.props.dataSourceManagement?.ui?.getDataSourceMenu<DataSourceViewConfig>();
      activeOption = this.state.dataSourceLoading
        ? undefined
        : [
          {
            label: this.state.dataSourceLabel,
            id: this.state.dataSourceId,
          },
        ];
    }
    const sideNav = [
      {
        name: Navigation.Notifications,
        id: 0,
        href: `#${Pathname.Channels}`,
        items: [
          {
            name: Navigation.Channels,
            id: 2,
            href: `#${Pathname.Channels}`,
            isSelected: pathname === Pathname.Channels,
          },
          {
            name: Navigation.EmailSenders,
            id: 3,
            href: `#${ROUTES.EMAIL_SENDERS}`,
            isSelected: pathname === ROUTES.EMAIL_SENDERS,
          },
          {
            name: Navigation.EmailGroups,
            id: 4,
            href: `#${ROUTES.EMAIL_GROUPS}`,
            isSelected: pathname === ROUTES.EMAIL_GROUPS,
          },
        ],
      },
    ];
    return (
      <CoreServicesConsumer>
        {(core: CoreStart | null) =>
          core && (
            <ServicesContext.Provider value={this.getServices(core.http)}>
              <ServicesConsumer>
                {(services: BrowserServices | null) =>
                  services && (
                    <MainContext.Provider value={this.state}>
                      <ModalProvider>
                        <DataSourceMenuContext.Provider
                          value={{
                            dataSourceId: this.state.dataSourceId,
                            dataSourceLabel: this.state.dataSourceLabel,
                            multiDataSourceEnabled: this.props.multiDataSourceEnabled,
                          }}
                        >
                          {this.props.multiDataSourceEnabled && DataSourceMenuView && DataSourceMenuSelectable && (
                            <Switch>
                              <Route
                                path={[
                                  `${ROUTES.EDIT_CHANNEL}/:id`,
                                  `${ROUTES.CHANNEL_DETAILS}/:id`,
                                  `${ROUTES.EDIT_SENDER}/:id`,
                                  `${ROUTES.EDIT_RECIPIENT_GROUP}/:id`,
                                  `${ROUTES.EDIT_SES_SENDER}/:id`
                                ]}
                                render={() => (
                                  <DataSourceMenuView
                                    setMenuMountPoint={this.props.setActionMenu}
                                    componentType={"DataSourceView"}
                                    componentConfig={{
                                      activeOption: [{ label: this.state.dataSourceLabel, id: this.state.dataSourceId }],
                                      dataSourceFilter: this.dataSourceFilterFn,
                                    }}
                                  />
                                )}
                              />
                              <Route
                                path={[
                                  "/",
                                  ROUTES.CHANNELS,
                                  ROUTES.CREATE_CHANNEL,
                                  ROUTES.CREATE_SENDER,
                                  ROUTES.CREATE_SES_SENDER,
                                  ROUTES.CREATE_RECIPIENT_GROUP,
                                  ROUTES.EMAIL_GROUPS,
                                  ROUTES.EMAIL_SENDERS,
                                  ROUTES.NOTIFICATIONS,
                                ]}
                                render={() => (
                                  <DataSourceMenuSelectable
                                    setMenuMountPoint={this.props.setActionMenu}
                                    componentType={"DataSourceSelectable"}
                                    componentConfig={{
                                      savedObjects: core?.savedObjects.client,
                                      notifications: core?.notifications,
                                      fullWidth: false,
                                      activeOption,
                                      onSelectedDataSources: this.onSelectedDataSources,
                                      dataSourceFilter: this.dataSourceFilterFn,
                                    }}
                                  />
                                )}
                              />
                              <Route
                                path={[ROUTES.CREATE_SES_SENDER, ROUTES.CREATE_CHANNEL, ROUTES.CREATE_RECIPIENT_GROUP, ROUTES.CREATE_SENDER]}
                                render={() =>
                                  this.state.dataSourceReadOnly ? (
                                    <DataSourceMenuView
                                      setMenuMountPoint={this.props.setActionMenu}
                                      componentType={"DataSourceView"}
                                      componentConfig={{
                                        activeOption: [{ label: this.state.dataSourceLabel, id: this.state.dataSourceId }],
                                        fullWidth: false,
                                        dataSourceFilter: this.dataSourceFilterFn,
                                      }}
                                    />
                                  ) : (
                                    <DataSourceMenuSelectable
                                      setMenuMountPoint={this.props.setActionMenu}
                                      componentType={"DataSourceSelectable"}
                                      componentConfig={{
                                        savedObjects: core?.savedObjects.client,
                                        notifications: core?.notifications,
                                        fullWidth: false,
                                        activeOption,
                                        onSelectedDataSources: this.onSelectedDataSources,
                                        dataSourceFilter: this.dataSourceFilterFn,
                                      }}
                                    />
                                  )
                                }
                              />
                            </Switch>
                          )}
                          <EuiPage>
                            {!this.state.dataSourceLoading && (
                              <>
                                <ModalRoot services={services} />
                                {pathname !== ROUTES.CREATE_CHANNEL &&
                                  !pathname.startsWith(ROUTES.EDIT_CHANNEL) &&
                                  !pathname.startsWith(ROUTES.CHANNEL_DETAILS) &&
                                  pathname !== ROUTES.CREATE_SENDER &&
                                  !pathname.startsWith(ROUTES.EDIT_SENDER) &&
                                  pathname !== ROUTES.CREATE_SES_SENDER &&
                                  !pathname.startsWith(ROUTES.EDIT_SES_SENDER) &&
                                  pathname !== ROUTES.CREATE_RECIPIENT_GROUP &&
                                  !pathname.startsWith(ROUTES.EDIT_RECIPIENT_GROUP) &&
                                  // Conditionally render sidebar based on the feature flag
                                  !core.chrome?.navGroup?.getNavGroupEnabled() && (
                                    <EuiPageSideBar style={{ minWidth: 155 }}>
                                      <EuiSideNav
                                        style={{ width: 155 }}
                                        items={sideNav}
                                      />
                                    </EuiPageSideBar>
                                  )}
                                <EuiPageBody>
                                  <Switch>
                                    <Route
                                      path={ROUTES.CREATE_CHANNEL}
                                      render={(props: RouteComponentProps) => (
                                        <CreateChannel {...props} />
                                      )}
                                    />
                                    <Route
                                      path={`${ROUTES.EDIT_CHANNEL}/:id`}
                                      render={(
                                        props: RouteComponentProps<{ id: string }>
                                      ) => <CreateChannel {...props} edit={true} />}
                                    />
                                    <Route
                                      path={`${ROUTES.CHANNEL_DETAILS}/:id`}
                                      render={(
                                        props: RouteComponentProps<{ id: string }>
                                      ) => <ChannelDetails {...props} />}
                                    />
                                    <Route
                                      path={ROUTES.CHANNELS}
                                      render={(props: RouteComponentProps) => (
                                        <Channels
                                          {...props}
                                          notificationService={
                                            services?.notificationService as NotificationService
                                          }
                                        />
                                      )}
                                    />
                                    <Route
                                      path={ROUTES.EMAIL_SENDERS}
                                      render={(props: RouteComponentProps) => (
                                        <EmailSenders
                                          {...props}
                                          notificationService={
                                            services?.notificationService as NotificationService
                                          }
                                        /> // send dataSourceId as props or externally
                                      )}
                                    />
                                    <Route
                                      path={ROUTES.EMAIL_GROUPS}
                                      render={(props: RouteComponentProps) => (
                                        <EmailGroups
                                          {...props}
                                          notificationService={
                                            services?.notificationService as NotificationService
                                          }
                                        />
                                      )}
                                    />
                                    <Route
                                      path={ROUTES.CREATE_SENDER}
                                      render={(props: RouteComponentProps) => (
                                        <CreateSender
                                          {...props}
                                        />
                                      )}
                                    />
                                    <Route
                                      path={`${ROUTES.EDIT_SENDER}/:id`}
                                      render={(props: RouteComponentProps) => (
                                        <CreateSender {...props} edit={true} />
                                      )}
                                    />
                                    <Route
                                      path={ROUTES.CREATE_SES_SENDER}
                                      render={(props: RouteComponentProps) => (
                                        <CreateSESSender
                                          {...props}
                                        />
                                      )}
                                    />
                                    <Route
                                      path={`${ROUTES.EDIT_SES_SENDER}/:id`}
                                      render={(props: RouteComponentProps) => (
                                        <CreateSESSender {...props} edit={true} />
                                      )}
                                    />
                                    <Route
                                      path={ROUTES.CREATE_RECIPIENT_GROUP}
                                      render={(props: RouteComponentProps) => (
                                        <CreateRecipientGroup
                                          {...props}
                                        />
                                      )}
                                    />
                                    <Route
                                      path={`${ROUTES.EDIT_RECIPIENT_GROUP}/:id`}
                                      render={(props: RouteComponentProps) => (
                                        <CreateRecipientGroup {...props} edit={true} />
                                      )}
                                    />
                                    <Redirect from="/" to={core.chrome?.navGroup?.getNavGroupEnabled() ? this.props.defaultRoute : ROUTES.CHANNELS} />
                                  </Switch>
                                </EuiPageBody></>
                            )}
                          </EuiPage>
                        </DataSourceMenuContext.Provider>
                      </ModalProvider>
                    </MainContext.Provider>
                  )
                }
              </ServicesConsumer>
            </ServicesContext.Provider>
          )
        }
      </CoreServicesConsumer>
    );
  }
}