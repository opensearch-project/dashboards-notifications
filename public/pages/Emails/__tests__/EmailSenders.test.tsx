/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { render, screen } from '@testing-library/react';
import React from 'react';
import { routerComponentPropsMock } from '../../../../test/mocks/routerPropsMock';
import {
  coreServicesMock,
  mainStateMock,
  notificationServiceMock,
} from '../../../../test/mocks/serviceMock';
import { CoreServicesContext } from '../../../components/coreServices';
import { ServicesContext } from '../../../services';
import { MainContext } from '../../Main/Main';
import { EmailSenders } from '../EmailSenders';

describe('<EmailSenders/> spec', () => {
  it('renders the component with SMTP config type', () => {
    const mainState = { ...mainStateMock,
      availableConfigTypes: [
        'slack',
        'chime',
        'microsoft_teams',
        'webhook',
        'email',
        'sns',
        'smtp_account',
        'ses_account',
        'email_group',
      ],
    };
    const routerComponentPropsMock = {
      // Mock other props as needed
      notificationService: notificationServiceMock,
    };
    const utils = render(
          <CoreServicesContext.Provider value={coreServicesMock}>
            <MainContext.Provider value={mainState}>
              <EmailSenders {...routerComponentPropsMock} />
            </MainContext.Provider>
          </CoreServicesContext.Provider>
    );
    expect(utils.container.firstChild).toMatchSnapshot();
    expect(screen.queryByText('SMTP senders')).not.toBeNull();
    expect(screen.queryByText('SES senders')).not.toBeNull();
  });

  it('renders the component without SMTP config type', async () => {
    const mainState = { ...mainStateMock,
      availableConfigTypes: [
        'slack',
        'chime',
        'microsoft_teams',
        'webhook',
        'email',
        'sns',
        'ses_account',
        'email_group',
      ],
    };
    const routerComponentPropsMock = {
      // Mock other props as needed
      notificationService: notificationServiceMock,
    };
    const utils = render(
          <CoreServicesContext.Provider value={coreServicesMock}>
            <MainContext.Provider value={mainState}>
              <EmailSenders {...routerComponentPropsMock} />
            </MainContext.Provider>
          </CoreServicesContext.Provider>
    );
    expect(utils.container.firstChild).toMatchSnapshot();
    expect(screen.queryByText('SMTP senders')).toBeNull();
    expect(screen.queryByText('SES senders')).not.toBeNull();
  });
});
