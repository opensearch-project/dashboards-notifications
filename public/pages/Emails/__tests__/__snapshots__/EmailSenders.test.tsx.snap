// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmailSenders/> spec renders the component with SMTP config type 1`] = `
<div
  class="euiText euiText--small"
>
  <h1>
    Email senders
  </h1>
</div>
`;

exports[`<EmailSenders/> spec renders the component without SMTP config type 1`] = `
<div
  class="euiText euiText--small"
>
  <h1>
    Email senders
  </h1>
</div>
`;
